File: thinking-orbs-mega/ports/svelte/src/lib/engine/ribbon.ts
import { fibDir, type InkColor, makeProj, paint, radiusScale } from "./core.js";
import type { Dot, ModeDraw } from "./types.js";

export const drawRibbon: ModeDraw = (
	ctx,
	size,
	t,
	dark,
	o,
	ink: InkColor | undefined,
) => {
	const center = size / 2;
	const radius = (size / 2) * 0.78;
	const spin = o.spin ?? 1;
	const cameraTilt = 0.3;
	const project = makeProj(t * 0.1 * spin, cameraTilt, center, center, 1);
	const radiusMultiplier = radiusScale(size, o.rsPow ?? 0.6);
	const dots: Dot[] = [];
	const ghostN = o.ghostN ?? 150;

	for (let i = 0; i < ghostN; i++) {
		const direction = fibDir(i, ghostN);
		const [x, y, z] = project(
			direction[0] * radius,
			direction[1] * radius,
			direction[2] * radius,
		);
		const depth = (z / radius + 1) / 2;
		dots.push({
			x,
			y,
			z,
			r: 0.8 * radiusMultiplier,
			white: 0.78,
			a: 0.1 + 0.22 * depth,
		});
	}

	const yaw = t * 0.24 * spin;
	const tilt = o.faceOn ? -cameraTilt : 0.55 + 0.3 * Math.sin(t * 0.18) * spin;
	const ux = Math.cos(yaw);
	const uy = 0;
	const uz = Math.sin(yaw);
	const vx = -uz * Math.sin(tilt);
	const vy = Math.cos(tilt);
	const vz = ux * Math.sin(tilt);
	const nx = uy * vz - uz * vy;
	const ny = uz * vx - ux * vz;
	const nz = ux * vy - uy * vx;
	const wobbleAmplitude = 0.23 * (o.wobMul ?? 1);
	const baseRadius = o.faceOn ? radius / (1 + 0.85 * wobbleAmplitude) : radius;
	const baseLanes = o.lanes ?? 5;
	const segments = o.segs ?? 88;
	const lanes = Math.max(1, Math.round(baseLanes * (o.bandMul ?? 1)));

	for (let lane = 0; lane < lanes; lane++) {
		const laneOffset = (lane - (lanes - 1) / 2) * 0.075;
		const edge =
			Math.abs(lane - (lanes - 1) / 2) / Math.max(1, (lanes - 1) / 2);
		for (let segment = 0; segment < segments; segment++) {
			const angle = (segment / segments) * 2 * Math.PI;
			const wobble =
				(0.16 * Math.sin(angle * 3 - t * 1.7 + lane * 0.22) +
					0.07 * Math.sin(angle * 5 + t * 1.1)) *
				(o.wobMul ?? 1);
			const radial = o.faceOn ? 1 + wobble : 1;
			const offset = o.faceOn ? laneOffset : laneOffset + wobble;
			const x = ux * Math.cos(angle) + vx * Math.sin(angle) + nx * offset;
			const y = uy * Math.cos(angle) + vy * Math.sin(angle) + ny * offset;
			const z = uz * Math.cos(angle) + vz * Math.sin(angle) + nz * offset;
			const length = Math.sqrt(x * x + y * y + z * z);
			const renderedRadius = baseRadius * radial;
			const [projectedX, projectedY, projectedZ] = project(
				(x / length) * renderedRadius,
				(y / length) * renderedRadius,
				(z / length) * renderedRadius,
			);
			const depth = (projectedZ / radius + 1) / 2;
			dots.push({
				x: projectedX,
				y: projectedY,
				z: projectedZ,
				r:
					((o.rBase ?? 1.1) + (o.rDepth ?? 1.7) * depth) *
					(1 - 0.25 * edge) *
					radiusMultiplier,
				white: 0.52 - 0.44 * depth + 0.18 * edge,
				a: 0.4 + 0.6 * depth,
			});
		}
	}
	paint(ctx, dots, dark, o.rMin, ink);
};

File: thinking-orbs-mega/ports/ios/ThinkingOrbs/Sources/ThinkingOrbs/Modes/Ribbon.swift
// Ribbon: an undulating sash of parallel strands rides a great circle —
// the "composing" state. The tuned preset freezes the 3D tumble
// (spin 0), leaving the traveling undulation on a fixed band.
//
// The same painter drives "breathing" via `faceOn`: a face-on circle
// whose radius — not its out-of-plane offset — undulates.

import CoreGraphics
import Foundation

let drawRibbon: ModeDraw = { ctx, size, t, dark, o, tint in
    let cx = size / 2
    let cy = size / 2
    let R = (size / 2) * 0.78
    // spin scales the 3D tumble; spin=0 freezes the band's orientation,
    // leaving only the traveling undulation
    let spin = o["spin"] ?? 1
    let faceOn = (o["faceOn"] ?? 0) != 0
    let cameraTilt = 0.3
    let pt = makeProj(yaw: t * 0.1 * spin, tilt: cameraTilt, cx: cx, cy: cy, scale: 1)
    let rs = radiusScale(size, o["rsPow"] ?? 0.6)

    var dots: [Dot] = []
    let ghostN = Int(o["ghostN"] ?? 150)
    for i in 0..<ghostN {
        let d = fibDir(i, ghostN)
        let (px, py, z) = pt(d.0 * R, d.1 * R, d.2 * R)
        let depth = (z / R + 1) / 2
        dots.append(Dot(x: px, y: py, z: z, r: 0.8 * rs, white: 0.78, a: 0.1 + 0.22 * depth))
    }

    // The band plane, precessing (frozen when spin=0). Face-on cancels
    // the camera tilt so the great circle projects as a true circle.
    let ya = t * 0.24 * spin
    let ta = faceOn
        ? -cameraTilt
        : 0.55 + 0.3 * sin(t * 0.18) * spin
    let ux = cos(ya)
    let uy = 0.0
    let uz = sin(ya)
    let vx = -uz * sin(ta)
    let vy = cos(ta)
    let vz = ux * sin(ta)
    // plane normal n = u × v
    let nx = uy * vz - uz * vy
    let ny = uz * vx - ux * vz
    let nz = ux * vy - uy * vx

    // Radial lobes swell past R, so pull the face-on base inward by most
    // of the amplitude. The silhouette stays framed as deformation grows.
    let wobbleAmplitude = 0.23 * (o["wobMul"] ?? 1)
    let baseRadius = faceOn
        ? R / (1 + 0.85 * wobbleAmplitude)
        : R

    let baseLanes = o["lanes"] ?? 5
    let segs = Int(o["segs"] ?? 88)
    let lanes = max(1, Int(jsRound(baseLanes * (o["bandMul"] ?? 1))))
    for w in 0..<lanes {
        let laneOff = (Double(w) - Double(lanes - 1) / 2) * 0.075
        let edge = abs(Double(w) - Double(lanes - 1) / 2) / max(1, Double(lanes - 1) / 2)
        for k in 0..<segs {
            let a = (Double(k) / Double(segs)) * 2 * .pi
            // the undulation: two traveling waves along the band; wobMul
            // scales the deformation — 0 is a clean band
            let wob = (0.16 * sin(a * 3 - t * 1.7 + Double(w) * 0.22)
                + 0.07 * sin(a * 5 + t * 1.1)) * (o["wobMul"] ?? 1)
            // Normal wobble becomes inward-only after sphere normalization.
            // Face-on instead varies the in-plane radius for visible lobes.
            let radial = faceOn ? 1 + wob : 1
            let off = faceOn ? laneOff : laneOff + wob
            let x = ux * cos(a) + vx * sin(a) + nx * off
            let y = uy * cos(a) + vy * sin(a) + ny * off
            let z = uz * cos(a) + vz * sin(a) + nz * off
            let l = (x * x + y * y + z * z).squareRoot()
            let radius = baseRadius * radial
            let (px, py, zr) = pt((x / l) * radius, (y / l) * radius, (z / l) * radius)
            let depth = (zr / R + 1) / 2
            dots.append(Dot(
                x: px, y: py, z: zr,
                r: ((o["rBase"] ?? 1.1) + (o["rDepth"] ?? 1.7) * depth) * (1 - 0.25 * edge) * rs,
                white: 0.52 - 0.44 * depth + 0.18 * edge,
                a: 0.4 + 0.6 * depth
            ))
        }
    }
    paint(ctx, &dots, dark: dark, rMin: o["rMin"] ?? 0.3, tint: tint)
}
